add-init-content (#6)
fhdsl/GDSCN_SARS_RStudio_on_AnVIL: 09-student_guide.Rmd

```{r, include = FALSE}
ottrpal::set_knitr_image_path()
```

# Student Activity Guide

This chapter contains the student instructions for the SARS-CoV-2 Phylogeny with RStudio activity.

## Introduction

> Since the beginning of the covid19 pandemic in December 2019, the world has experienced waves of increased cases caused by new variants. This activity walks you through a simple phylogeny to explore how the new variants are related.

Reviewer: ```suggestion
Since the beginning of the Covid-19 pandemic in December 2019, the world has experienced waves of increased cases caused by new variants. This activity walks you through a simple phylogeny to explore how the new variants are related.
```

Reviewer: Could give a shout out to https://nextstrain.org/ - created by folks at Hutch!

### Before You Start

> Make sure you have an AnVIL account (refer to https://jhudatascience.org/AnVIL_Book_Instructor_Guide/student-account-setup.html) 

Reviewer: A great spot to use cow::borrow_chapter if you are so inclined :)

### Objectives

This activity will teach you how to use the AnVIL platform to:

1. Get started working on AnVIL
1. Launch RStudio
1. Import data into RStudio
1. Examine fasta and phyDat files
1. Build a neighbor-joining phylogeny
1. Interpret the topology and branch lengths of a phylogeny

## Getting Started

In the next few steps, you will walk through how to get set up to use RStudio on the AnVIL platform. AnVIL is centered around different “Workspaces”. Each Workspace functions almost like a mini code laboratory - it is a place where data can be examined, stored, and analyzed. The first thing we want to do is to copy or “clone” a Workspace to create a space for you to experiment.

Use a web browser to go to the AnVIL website. In the browser type:

```
anvil.terra.bio
```

:::{.notice}
**Tip**
At this point, it might make things easier to open up a new window in your browser and split your screen. That way, you can follow along with this guide on one side and execute the steps on the other.
:::

> ### Set Up

Reviewer: I'm guessing RStudio specific set up will go here?

Author: Yup, I was waiting until the child .rmd files were ready in the template to save work for myself

## Exercise One: Loading libraries in RStudio

Before we can start our analysis of how the SARS-CoV-2 variants are related to each other, we need to prepare the RStudio workspace and load the data. 

R is an open-source statistical programming language and anyone can contribute to it. People have written programs in R to do a ton of different things, and they can make those programs (known as packages, or libraries) available to everyone. Generally, when someone has created a package they want to share, they will submit it to a repository, where anyone using R can download it. 

> For this lesson, we are using the CRAN (Comprehensive R Archive Network) repository. There are a series of servers around the world that store the up-to-date packages. When you open R, you can access those servers and download any package you want. If we are downloading a package that has been stored on CRAN, we use the command "install.packages".

Reviewer: ```suggestion
For this lesson, we are using the [CRAN](https://cran.r-project.org/) (Comprehensive R Archive Network) repository. There are a series of servers around the world that store the up-to-date packages. When you open R, you can access those servers and download any package you want. If we are downloading a package that has been stored on CRAN, we use the command `install.packages`.
```

We need to install two packages:
[ape](https://academic.oup.com/bioinformatics/article/20/2/289/204981) and [phangorn](https://academic.oup.com/bioinformatics/article/27/4/592/198887). Both of these packages were written specifically for phylogenetic analysis in R.

To install the packages, we type the following code into the RStudio console:

```{r, message = FALSE, warning = FALSE}
install.packages('ape')

install.packages('phangorn')
```

> Once you've downloaded a package, it will be saved on your computer (or, in the case of AnVIL, on your persistent disk space) so that you don't have to download it again. Anytime you want to use the set of commands that are stored in a particular package, you'll tell R to open the package with the "library" command. 

Reviewer: ```suggestion
Once you've downloaded a package, it will be saved on your computer (or, in the case of AnVIL, on your persistent disk space) so that you don't have to download it again. Anytime you want to use the set of commands that are stored in a particular package, you'll tell R to open the package with the `library` command. 
```

<img src="resources/images/09-fasta.png" title="Examples of a fasta file format" alt="An example of the first few lines of a fasta file" style="display: block; margin: auto;" />

We can load this file into RStudio using the `read.FASTA` command and save it as the object "spike.small".

```{r, message = FALSE, warning = FALSE}
spike.fasta <- read.FASTA("sars_spike_protein.fasta")
```

After we've created an object in RStudio, we can get information about the object by typing the object's name.

```{r, message = FALSE, warning = FALSE}
spike.fasta
```

Here we can see a summary of what this object contains, as well as how long the sequences are and the sequence names. 

Notice that RStudio has saved the information in the fasta file as binary data. This means the sequence information has been converted from "ATCG" into something easier for RStudio to work with, but harder for humans to interpret.

The `phangorn` package uses a special data format called `phyDat`, which is derived from the fasta format. A `phyDat` object provides some additional information about the samples we upload. 

```{r, message = FALSE, warning = FALSE}

> spike.phydat <- read.phyDat("sars_spike_protein.fasta", format = "fasta")

Reviewer: will this work even with a cloned workspace?

Author: I'm not sure I understand the question.

spike.phydat
```

Different site patterns refers to sites that differ between sequences. In this small dataset, 46 of the 3827 possible bases (characters) show differences among these SARS-CoV-2 samples.

We've been working with a dataset that contains the original SARS-CoV-2 sequence (the Wuhan reference sample), as well as samples of the alpha, beta, delta, and gamma variants. Another variant, the omicron variant, was first identified in late 2021 and quickly became a variant of concern. Let's look at a dataset that contains additional omicron samples.

```{r, message = FALSE, warning = FALSE}
spike.omicron <- read.phyDat("sars_spike_protein_omicron.fasta", format = "fasta")

spike.omicron
```

::: {.reflection}
QUESTIONS:

1. What is some information saved in a .fasta object that RStudio tells us that we don't get from a .phyDat object?

2. How many omicron sequences are there in the second file (loaded into the object `spike.omicron`)?

3. Do you think there is more variability in the omicron sequences than in other variants (alpha, beta, delta, and gamma)? Why or why not?
:::

## Exercise Three: How are variants related to the original SARS-CoV-2 reference strain?

> We'll work with both the smaller spike protein dataset (saved as  `spike.small`) and the spike protein dataset with omicron samples (saved as `spike.omicron`) in this activity. 

Reviewer: isn't it saved as `spike.phydat`? This could be a tad confusing

Author: Bah, thank you for catching this!

### Building a neighbor-joining tree

We can use `phyDat` objects to build a type of phylogenetic tree called a neighbor-joining tree. We can do this in two steps in RStudio using the `phangorn` package.

```{r, message = FALSE, warning = FALSE}
dist.small <- dist.ml(spike.phydat)
nj.small <- nj(dist.small)
```

The first command (`dist.ml`) converts the genetic data into a distance matrix, while the second command (`nj`) uses the matrix to group taxa based on the genetic distance between them. We plot the object created by the `nj` command to visualize our phylogenetic tree.

```{r, message = FALSE, warning = FALSE}
plot(nj.small)
```

We can do the same for the larger dataset that includes omicron sequences.

```{r, message = FALSE, warning = FALSE}
dist.omicron <- dist.ml(spike.omicron)
nj.omicron <- nj(dist.omicron)
plot(nj.omicron)
```

> When reading phylogenetic trees, it's important to remember that relatedness is determined by the number of nodes between two taxa. The delta variant is most closely related to the beta variant because they share a common node (where the branches join together). We always look at common ancestors of clades, so we essentially read ancestry on a tree from right (where the taxa are) to left. Thus, the delta variant is more closely related to the beta variant because the delta branch connects to a node shared by the beta branch before it connects to a node shared with the gamma branch or the Wuhan reference/alpha variant branch.

Reviewer: What if you created an image where there are colored shapes on the nodes you are talking about, for reference? Could just do this in powerpoint and use ottrpal, nothing crazy with ggplot :)

Author: Ooooh, good idea!

First, let's load and view a tree containing the five non-omicron sequences.

```{r, message = FALSE, warning = FALSE}
spike.tree_small <- read.tree("sars_tree_small.tre")

plot(spike.tree_small)
edgelabels(spike.tree_small$edge.length)
```

The numbers in green boxes represent the number of DNA changes along a particular branch. For example, the Wuhan reference sequence is estimated to have only one DNA change from the spike protein sequence of the most recent common ancestor of all SARS-CoV-2 strains, while the alpha variant has 6 bases different from the most recent common ancestor. This also means the distance between Wuhan reference sequence and the alpha sample is 7 substitutions (1 substitution for the Wuhan sample + 6 substitutions for the alpha variant sample).

::: {.dictionary}
**Breakout Box: Learn more about phylogenetically informative sites**

You may be wondering why the total branch length changes (31) is different from the number of site patterns listed in the phyDat summary (46). This is because only some of the 46 site patterns are what we call _phylogenetically informative_.

Let's look at a short example:

> cat     GC*A* TT*C*

Reviewer: Need to use `** **` if you want bold

:::

Now let's load and view a tree containing all nine SARS-CoV-2 samples.

```{r, message = FALSE, warning = FALSE}
spike.tree_full <- read.tree("sars_tree_full.tre")

plot(spike.tree_full)
edgelabels(spike.tree_full$edge.length)
```

What do these branch lengths tell us about the relationship among the SARS-CoV-2 variants?

::: {.reflection}
QUESTIONS:

10. Did adding the omicron samples change the branch length distances between the original five samples? What is the branch length distance between the delta and beta variants now?

11. What is the branch length distance between the omicron clade and its closest relative?

12. The covid vaccines were originally designed based on the Wuhan reference sequence of the spike protein. Using what you have learned about the phylogenetic tree of SARS-CoV-2 variants, can you explain why these initial vaccines were less effective at protecting against the omicron variants than they were against the delta variant?
:::

## Exercise Four: Do other protein-coding regions show us the same phylogenetic relationships?

> The SARS-CoV-2 genome contains 6 protein-coding regions. So far, we've been working with genetic data from the region that codes for spike protein. (The spike protein is the part of the virus that sticks out to form the characteristic spikes on the outside of the SARS-CoV-2 virus.) However, you could also use any of the other 5 protein-coding regions for this exercise. 

Reviewer: Could be cool to have a picture of the virus and point to the spike protein when it's first mentioned above

Author: YES
